### PR TITLE
fix: #12253 Anchor picker does not appear

### PR DESCRIPTION
**Pull request by marcoteodoro**
This relates to #12253 

### Steps to test 

1. Install Umbraco v11 
2. Use database attached [Umbraco.sqlite.zip](https://github.com/umbraco/Umbraco-CMS/files/10105668/Umbraco.sqlite.zip)
3. Open umbraco with u/n : marcoteodoro@double.pt p/w: 0123456789
4. Open destination node and create a new anchor link from the origin node. (you will see that anchor links from the blocklist editor and grid editor are listed on the dropdown)

chrome-capture-2022-10-28

### Description

`GetAnchorValuesFromRTEs` method was updated to include `BlockList` and `BlockGrid`
https://github.com/umbraco/Umbraco-CMS/blob/c9410391bf5cc7a05c138208fbbbb404e85847bf/src/Umbraco.Core/Services/ContentServiceExtensions.cs#L80-L80

on top of this, i had to include this line of code `var rteUnescaped = Regex.Unescape(rteContent);` to make the regex work. https://github.com/umbraco/Umbraco-CMS/blob/c9410391bf5cc7a05c138208fbbbb404e85847bf/src/Umbraco.Core/Services/ContentServiceExtensions.cs#L103-L103

However, one thing i noticed is that this that when the culture variable is assigned with `*` this method is always returning null contentProperty.GetValue(culture)?.ToString();`

https://github.com/umbraco/Umbraco-CMS/blob/c9410391bf5cc7a05c138208fbbbb404e85847bf/src/Umbraco.Core/Services/ContentServiceExtensions.cs#L73

So this PR is just an initial commit to see what others can suggest. Is this last issue a known issue? or I'm not using it correctly?

### CODE REVIEW

**Comment by lauraneto** [comment]
Olá @marcoteodoro , 🙂

Thank you for your contribution! And really sorry for the long delay. 😣

I reviewed and tested your pull request and I noticed that your solution only works when you have a block list (same for block grid) property in your document that directly contains a block with an RTE inside, it does not work when you have more nesting (a block inside a block).
This is because you are only unescaping the RTE content once and for each layer of nesting the content gets more and more serialized.

While this is already a nice improvement, I wonder, what do you think about instead of unescaping the content, just adjusting the regex instead?
For example, if we change it to something like `<a id=\\*\"(.*?)\\*\">`, it should allow for any nesting level to be matched.

About your question regarding the culture, with your database and my own, I never got it to be `*` when using the link picker UI. (`GetUrlAndAnchors` always assigns a culture or null, if it would be `*`).
If you have reproduction steps do let me know so I can test, but either way, as this is a public method that can be called with `*` I think the fix is good nonetheless.

File: src/Umbraco.Core/Services/ContentServiceExtensions.cs [hunk 1]
```diff
@@ -67,20 +68,25 @@ public static IContent CreateContent(this IContentService contentService, string
     public static IEnumerable<string> GetAnchorValuesFromRTEs(this IContentService contentService, int id, string? culture = "*")
     {
         var result = new List<string>();
+
+        //TODO : remove this hack : contentProperty.GetValue(culture) always return null when culture variable is "*"
+        if (culture == "*")
+        {
+            culture = string.Empty;
+        }
```
Comment (lauraneto) [inline src/Umbraco.Core/Services/ContentServiceExtensions.cs:76]:
```suggestion
        culture = culture is not "*" ? culture : null;
```
Looking at similar code, it seems like normally it should just default to `null`.

File: src/Umbraco.Core/Services/ContentServiceExtensions.cs [hunk 2]
```diff
@@ -93,7 +99,8 @@ public static IEnumerable<string> GetAnchorValuesFromRTEs(this IContentService c
         string rteContent)
     {
         var result = new List<string>();
-        MatchCollection matches = AnchorRegex.Matches(rteContent);
+        var rteUnescaped = Regex.Unescape(rteContent);
+        MatchCollection matches = AnchorRegex.Matches(rteUnescaped);
         foreach (Match match in matches)
         {
             result.Add(match.Value.Split(Constants.CharArrays.DoubleQuote)[1]);
```
Comment (lauraneto) [inline src/Umbraco.Core/Services/ContentServiceExtensions.cs:106]:
```suggestion
            result.Add(match.Groups[1].Value);
```
No need to be manipulating strings when we already have the capture group and this change is required for the new regex.

File: src/Umbraco.Core/Services/ContentServiceExtensions.cs [hunk 3]
```diff
@@ -93,7 +99,8 @@ public static IEnumerable<string> GetAnchorValuesFromRTEs(this IContentService c
         string rteContent)
     {
         var result = new List<string>();
-        MatchCollection matches = AnchorRegex.Matches(rteContent);
+        var rteUnescaped = Regex.Unescape(rteContent);
+        MatchCollection matches = AnchorRegex.Matches(rteUnescaped);
```
Comment (lauraneto) [inline src/Umbraco.Core/Services/ContentServiceExtensions.cs:103]:
```suggestion
        MatchCollection matches = AnchorRegex.Matches(rteContent);
```

File: src/Umbraco.Core/Services/ContentServiceExtensions.cs [hunk 4]
```diff
@@ -1,8 +1,9 @@
-﻿// Copyright (c) Umbraco.
+// Copyright (c) Umbraco.
 // See LICENSE for more details.
 
 using System.Text.RegularExpressions;
 using Umbraco.Cms.Core;
+using Umbraco.Cms.Core.HealthChecks;
```
Comment (lauraneto) [inline src/Umbraco.Core/Services/ContentServiceExtensions.cs:6]:
```suggestion
```

**Review by lauraneto — CHANGES_REQUESTED** [review]
Unfortunately I can't suggest an edit on the regex (GitHub doesn't allow me to suggest edits in unchanged code 😅).
But if you do the suggested changes and update the regex as per my other comment it should do the trick. But please verify! 🙂

**Comment by abjerner** [comment]
Maybe not for this PR, but in general it might be worth looking at the anchor picker from another angle. Right now it's very much based on RTEs as emphasized by the method name: `GetAnchorValuesFromRTEs`.

I'd say their could be plenty of other things that could be used as an anchor. Eg. when using the block list, each block item could potentially be an anchor. That's a use case that I've recently had where editors want to deep link to specific block elements. I don't think the anchor picker has been updated to properly support the grid and now the block list, although it probably should have been.

There really isn't a good way to support this with the current implementation. The `GetAnchorValuesFromRTEs` method is also static meaning developers don't really have a way to override the default behavior.

So maybe an `IAnchorProvider` interface could be introduced, and used instead. The default implementation could be that of the existing `GetAnchorValuesFromRTEs` or Marco's proposed changes, but if set up correctly with dependency injection, developers can override the default implementation.

**Comment by lauraneto** [comment]
Hi @abjerner ,
I totally see your use case (as I also have that same scenario on most of my projects 😅).
But yes, might be best to create a separate discussion/issue regarding this, so we can work on it on top of this pull request (and maybe discuss if there is a solution that we can build within Umbraco or if indeed an interface would be the way to go, so at least we (umbraco devs) can have control over the behavior if we want).

**Review by lauraneto — APPROVED** [review]

**Comment by lauraneto** [comment]
Hi @marcoteodoro ,

I have just merged this pull request!
Thanks again for the contribution and it was nice to see you in Codegarden (and also see you finish this one during the hackathon 🙂).

I do see that this is your first merged PR, so many congrats!
If you have an account on [our.umbraco.com ](https://our.umbraco.com/) then drop a comment below with your account name and @nul800sebastiaan will add a shiny new contributor badge to your profile. 🏆

Thanks again, and looking forward to your next PR... Or seeing you in the next Codegarden, maybe? 😄

**Comment by marcoteodoro** [comment]
Hi @lauraneto,

I am thrilled to see my first pull request merged, and I am very grateful for your support throughout the process. The experience at Codegarden was indeed an enriching one, and having the opportunity to complete this PR during the hackathon was a milestone I will always cherish.

Your guidance and positive feedback have inspired me to contribute more, and I look forward to the journey ahead with great excitement. This feels like just the beginning, and I am eager to keep learning, growing, and positively impacting our community. 

Here's my profile link on our.umbraco.com: https://our.umbraco.com/members/marcoteodoro/. I look forward to seeing that shiny new contributor badge on my profile, @nul800sebastiaan! 🏆

I'm already counting down the days until the next Codegarden. It would be wonderful to see you and the team there again, and who knows, maybe even present another pull request or two. 😉

Once again, thank you for everything. Here's to many more collaborative experiences and shared successes in the future. Cheers!

See you all next year!

**Comment by nul800sebastiaan** [comment]
Whoop whoop @marcoteodoro! Pleasure to have you at the hackathon and as a contributor! 🎉 You have your contrib badge to show off now and some more good news, this fix will also be available in v10 and 11, cherry picked in 2f35967bb28ca118a443580209e74fac42a682a3

**Comment by marcoteodoro** [comment]
Many thanks, @sebastian-claesson! and once again @lauraneto! I'm I right to think this will be released on the next v10 release? and if so when is that going to happen?

**Pull request merged**